plugins/hls-selection-range-plugin/src/Ide/Plugin/SelectionRange/CodeRange.hs

```diff
@@ -0,0 +1,109 @@
+import           Development.IDE.GHC.Compat.Util
+import           Development.IDE.GHC.ExactPrint          (GetAnnotatedParsedSource (GetAnnotatedParsedSource))
+import           GHC.Generics                            (Generic)
+import           Ide.Plugin.SelectionRange.ASTPreProcess (PreProcessEnv (..),
+                                                          preProcessAST)
+import           Prelude                                 hiding (log)
+
+data Log = LogShake Shake.Log
+    | LogBadDependency BadDependencyLog
+    | LogNoAST
+
+instance Pretty Log where
+    pretty log = case log of
+        LogShake shakeLog                 -> pretty shakeLog
+        LogBadDependency badDependencyLog -> pretty badDependencyLog
+        LogNoAST                          -> "no HieAst exist for file"
+
+data BadDependencyLog = forall rule. Show rule => BadDependencyLog rule
+
+instance Pretty BadDependencyLog where
+    pretty (BadDependencyLog rule) = "can not get result from rule " <> pretty (show rule)
+
+data CodeRange = CodeRange Range [CodeRange]
+    deriving (Show, Generic)
+
+instance NFData CodeRange
+
+buildCodeRange :: HieAST a -> RefMap a -> Annotated ParsedSource -> CodeRange
+buildCodeRange ast refMap _ =
+    -- We work on 'HieAST', then convert it to 'CodeRange', so that applications such as selection range and folding
+    -- range don't need to care about 'HieAST'
+    -- TODO @sloorush actually use 'Annotated ParsedSource' to handle structures not in 'HieAST' properly (for example comments)
+    let ast' = runReader (preProcessAST ast) (PreProcessEnv refMap)
+    in simplify . astToCodeRange $ ast'
+
+astToCodeRange :: HieAST a -> CodeRange
+astToCodeRange (Node _ sp []) = CodeRange (realSrcSpanToRange sp) []
+astToCodeRange (Node _ sp children) = CodeRange (realSrcSpanToRange sp) (fmap astToCodeRange children)
+
+-- Remove redundant nodes in 'CodeRange' tree
+simplify :: CodeRange -> CodeRange
+simplify r@(CodeRange range1 [CodeRange range2 children])
+    | range1 == range2 = CodeRange range1 children
+    | otherwise = r
+simplify r = r
+
+data GetCodeRange = GetCodeRange
+    deriving (Eq, Show, Typeable, Generic)
+
+instance Hashable GetCodeRange
+instance NFData   GetCodeRange
+
+type instance RuleResult GetCodeRange = CodeRange
+
+-- | Like use, but report absense in 'ExceptT'
+useExcept :: IdeRule k v => (BadDependencyLog -> msg) -> k -> NormalizedFilePath -> ExceptT msg Action v
+useExcept f rule = maybeToExceptT (f (BadDependencyLog rule)) . MaybeT . use rule
+
+codeRangeRule :: Recorder (WithPriority Log) -> Rules ()
+codeRangeRule recorder =
+    define (cmapWithPrio LogShake recorder) $ \GetCodeRange file -> handleError recorder $ do
+        -- We need both 'HieAST' (for basic AST) and api annotations (for comments and some keywords).
+        -- See https://gitlab.haskell.org/ghc/ghc/-/wikis/api-annotations
+        HAR{hieAst, refMap} <- useExcept LogBadDependency GetHieAst file
```

Comment:
This seems suspicious. I thought BadDependency was special, and we shouldn't be catching it like this?

Reply:
To clarify, `LogBadDependency` is just a log type constructor. It has no relation with `BadDependency` exception in  ghcide.

Do we prefer throwing an exception in this case? (indirectly through `use_` or `uses_`?)

I can't find too many usages of `use_`, `uses_` and `BadDependency` in plugins. It seems that most code just handles the `Maybe` manually, and returns an empty result on `Nothing`.

For example:
https://github.com/haskell/haskell-language-server/blob/510bd51e46fea8fb51ddfaa60bba505f0663497d/plugins/hls-hlint-plugin/src/Ide/Plugin/Hlint.hs#L371-L375

https://github.com/haskell/haskell-language-server/blob/510bd51e46fea8fb51ddfaa60bba505f0663497d/plugins/hls-hlint-plugin/src/Ide/Plugin/Hlint.hs#L399-L408

Reply:
Oh I see, I misunderstood. I think it's fine to log like you're doing here. I agree it's pretty inconsistent how people use these things (also of interest to @drsooch I think).

Reply:
And maybe @July541

Reply:
In fact, I'm a bit unsure too. In the latest commit, I changed it to `use_`. Maybe we should discuss this, and:

1. If we decide to use exceptions, then we should advocate `use_` and `uses_`, and my previous ExceptT trick with a `BadDependency` log type is no longer necessary.
2. If we decide not to use exceptions, maybe we could mark `use_` and `uses_` as deprecated, and add a log type for `BadDependency` directly in `Shake.hs`.